data/blog/umdctf-2023/pokeptcha.md

```diff
@@ -0,0 +1,414 @@
+    values[moWJn] = parseInt(data_array[i]) ^ 21
+  }
+  return values
+}
+```
+
+So this gives us the values we want, just split at the pipes, parseint and then xor with 21. Next there were some utility functions to read parts of the values array (which is assigned to a global value called `bytecode`):
+
+```javascript
+function next_index() {
+  pc = pc + 1
+  return pc - 1
+}
+
+function read_string() {
+  var l = bytecode[next_index()]
+  var s = ''
+  for (let i = 0; i < l; i = i + 1) {
+    s = s + String.fromCharCode(bytecode[next_index()])
+  }
+  return s
+}
+
+function read_const() {
+  var currentIndex = bytecode[next_index()]
+  if (currentIndex & 1) {
+    return read_string()
+  } else {
+    return bytecode[next_index()]
+  }
+}
+
+function read_single() {
+  return global_data[next_index()]
+}
+```
+
+We see that this just reads from the `bytecode` array.
+
+Next we have some more utility functions to push and pop to a global `stack` variable:
+
+```javascript
+function pop() {
+  var length = stack.length
+  var popped_val = stack[length - 1]
+  stack.length = length - 1
+  return popped_val
+}
+
+function push(val) {
+  var sp = actual_stack.length
+  stack[sp] = val
+}
+```
+
+Then we have an array of three functions (Which are each assigned individually, but in the end we have `operations = [push_func, eval_func, assign_func]`), which are defined as follows:
+
+```javascript
+// 0
+function push_func() {
+  stack.push(read_const())
+}
+
+// 1
+function eval_func() {
+  push(eval(pop()))
+}
+
+// 2
+function assign_val(value, key, obj) {
+  obj[key] = value
+}
+function assign_func() {
+  var values = []
+  for (let i = 0; i < 3; i = i + 1) {
+    values.push(pop())
+  }
+  assign_val(...values)
+}
+```
+
+Where the comment indicates the position in the array. The final piece of this whole thing is:
+
+```javascript
+function main_loop(bone_in) {
+  reset()
+  stack.push(bone_in)
+  // QZOR is the start time, will be used in the bytecode
+  QZOR = Date.now()
+
+  // while(true)?
+  for (; 1 < 100; ) {
+    var op = read_single()
+
+    // Special handler for opcode 100 -> exit
+    if (op == 100) {
+      memory = []
+      // LZX will be set in the bytecode
+      return LZX
+    } else {
+      var func = operations[op]
+      // Don't really understand this...
+      if (func == 2) {
+        break
+      }
+      // execute op
+      func()
+    }
+  }
+  reset()
+}
+```
+
+Which, in case my variable names have not made it clear yet, should hopefully make it clear that we’re dealing with a VM in JavaScript. We take one single value as opcode and then execute the operation at that index (The comment above the function). However we currently only have indices 0-2 defined, but looking at the values there is no way that can be true. At the same time we see some code in the array, so I wrote a small interpreter in python which steps through and prints when we assign to anything. The first thing we do is
```

Comment:
```suggestion
Which, in case my variable names have not made clear yet, should indicate that we’re dealing with a VM in JavaScript. We take a single value as an opcode, and then execute the operation at that index (The comment above the function). However, we currently only have indices 0-2 defined, and looking at the values there is no way that can be true. At the same time we see some code in the array, so I wrote a small interpreter in Python which steps through and prints when we assign to anything. The first thing we do is:
```